### PR TITLE
[SMTChecker] Move CHC counterexamples to primary location

### PR DESCRIPTION
**Pull request by leonardoalt**
This especially enables testing CHC counterexamples. This is really needed now that they're becoming more complex.
Now we just need to review these updates :p

Depends on https://github.com/ethereum/solc-js/pull/506

### CODE REVIEW

**Review by axic — DISMISSED** [review]

**Comment by leonardoalt** [comment]
I think we can merge this without checking the results, and then check them separately and open issues for buggy counterexamples.

File: libsolidity/formal/CHC.cpp
Comment (axic) [inline libsolidity/formal/CHC.cpp:1390]: As @leonardoalt explained the counterexample has no specific source location, but may as well be tied to this location. Hence I'm fine with the change.

However I think it would be nice to split the single message field in `Error` into two: "message" and "additionalDetails". In this example the counterexample goes into the additional detail.

The motivation: IDEs wouldn't need to introduce their means to truncate long messages.

Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:1390]: Agree with that.

**Comment by leonardoalt** [comment]
1 nondeterminism case on Ubuntu
9 nondeterminism cases on OSX
The Ubuntu one we could just disable for now, as we did with other tests, but the OSX ones not sure...

File: libsolidity/formal/CHC.cpp
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:1390]: @axic do you mean this should be done in this PR? Or future?

**Comment by leonardoalt** [comment]
@ekpyron @axic @chriseth We need to decide whether the counterexamples should be tested or not.
It would be nice to test them mostly to catch problems in the counterexamples caused by other features.
The downside is that they're a bit more nondeterministic than queries, as seen above.
Do we want to go for it or not?

**Comment by ekpyron** [comment]
> @ekpyron @axic @chriseth We need to decide whether the counterexamples should be tested or not.
> It would be nice to test them mostly to catch problems in the counterexamples caused by other features.
> The downside is that they're a bit more nondeterministic than queries, as seen above.
> Do we want to go for it or not?

At least I'd say your call - you're who primarily will need to deal with the nondeterminism :-). I'm pretty sure it will be at least somewhat annoying... What you could also consider as compromise would be having explicit support in ``SMTCheckerTest``, i.e. adding a test configuration option to filter them out or not on a per-test basis. Then you can test some stable ones to make sure the counterexample mechanism doesn't start reporting nonsense - but you can disable flaky ones... (while not having to remove the entire tests for it)

**Comment by chriseth** [comment]
I agree, we can add some mechanism so that running the tests in different environments still works.

**Comment by leonardoalt** [comment]
Added an option `SMTIgnoreCex` to ignore SMTChecker counterexamples.
That was a good idea @ekpyron @chriseth 
Had to ignore a bunch of flaky ones, but the rest seems to work.
This also depends on https://github.com/ethereum/solc-js/pull/506

File: libsolidity/formal/CHC.cpp
Comment (ekpyron) [reply libsolidity/formal/CHC.cpp:1390]: I'd vote for "future" - for changing the error reporting mechanism like that I'd expect quite some work to be done on the error reporter side that's too independent of this change to cram into this PR.

File: test/libsolidity/SMTCheckerTest.cpp
Comment (ekpyron) [inline test/libsolidity/SMTCheckerTest.cpp]:
```suggestion
				e.message = e.message.substr(0, cexPos);
```
maybe? Not too important.

File: test/libsolidity/SMTCheckerTest.cpp
Comment (ekpyron) [inline test/libsolidity/SMTCheckerTest.cpp]:
```suggestion
		removeCex(m_errorList);
```

We can assume no counterexample in the expectations in these cases, can't we?

File: test/libsolidity/smtCheckerTests/array_members/push_2d_arg_1_unsafe.sol
Comment (ekpyron) [inline test/libsolidity/smtCheckerTests/array_members/push_2d_arg_1_unsafe.sol]: Ah, maybe we can't assume no counterexamples in the expectations... but wouldn't it be better if there weren't any? Otherwise it looks like they're tested, but they aren't... is it hard to just filter the counterexample when ``isoltest`` updates expectations as well?

**Review by ekpyron — COMMENTED** [review]
Some comments, but if there's issues with doing things differently, I'd be fine with keeping things as they are now.

File: test/libsolidity/SMTCheckerTest.cpp
Comment (leonardoalt) [reply test/libsolidity/SMTCheckerTest.cpp]: That's the thing, I would rather allow the test file to have expectations which are generated by whoever wrote the PR locally, and we just ignore them in the test.

File: test/libsolidity/smtCheckerTests/array_members/push_2d_arg_1_unsafe.sol
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/array_members/push_2d_arg_1_unsafe.sol]: Well it's quite easy to also see that they are ignored, if the option `SMTIgnore: yes` is there. As said in the other comment, I would also rather have them in the file and just ignore in the comparison.

File: test/libsolidity/SMTCheckerTest.cpp
Comment (leonardoalt) [reply test/libsolidity/SMTCheckerTest.cpp]: yea ofc

**Comment by ekpyron** [comment]
Ok, I see it's intentional, but I'm not sure I understand why... I would argue the test expectations should better not contain anything that's not actually being tested, so I'd cut them out.

> That's the thing, I would rather allow the test file to have expectations which are generated by whoever wrote the PR locally.

What's the value in that?

I think I'd be fine to do it that way in the end, but I'd be curious about the reasoning behind it :-).

**Review by ekpyron — DISMISSED** [review]
So since it's not that big a deal to me, I'm already approving this - but I'm still wondering what's the value in having unchecked counterexamples in the expectations and I myself would rather remove them from there...

**Comment by leonardoalt** [comment]
For me the value is that when I write a PR and get counterexamples while running isoltest, I can still check the counterexample myself, even if CI doesn't.

**Comment by ekpyron** [comment]
> For me the value is that when I write a PR and get counterexamples while running isoltest, I can still check the counterexample myself, even if CI doesn't.

Yeah, ok... it might actually also make reviewing test cases easier to have counterexamples in them...
I'm still not entirely sure, because it still is weird to have them there, but untested, but I'm fine with it :-).

**Comment by leonardoalt** [comment]
Also fine removing it, not a big deal to do it.

**Comment by ekpyron** [comment]
> Also fine removing it, not a big deal to do it.

Then throw some dice ;-). No really, as you like then!

**Comment by ekpyron** [comment]
Or anyone else leaning either way?

**Comment by leonardoalt** [comment]
Now I already removed them locally haha

**Review by ekpyron — DISMISSED** [review]
Reapproving.

**Comment by leonardoalt** [comment]
I'll still need to change the solc-js branch after it gets merged there.

File: libsolidity/formal/CHC.cpp
Comment (axic) [reply libsolidity/formal/CHC.cpp:1390]: Can you create an issue an leave a TODO item of said issue in this piece of code?

Comment (axic) [reply libsolidity/formal/CHC.cpp:1390]: I actually wonder if tooling (Truffle, Remix, specifically) displays secondary source locations? Or do they just use the formatted message in standard json?

File: test/libsolidity/SMTCheckerTest.cpp
Comment (leonardoalt) [inline test/libsolidity/SMTCheckerTest.cpp:57]: @axic this option did not exist before, it is introduced into the test framework in this PR

File: libsolidity/formal/CHC.cpp
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:1390]: https://github.com/ethereum/solidity/issues/10534

File: test/externalTests/solc-js/solc-js.sh
Comment (axic) [inline test/externalTests/solc-js/solc-js.sh]: Fix this please.

Comment (leonardoalt) [reply test/externalTests/solc-js/solc-js.sh]: Nice!

**Review by ekpyron — APPROVED** [review]
For the record: I didn't sanity-check the actual counterexamples :-). But fine to merge.

**Comment by leonardoalt** [comment]
> For the record: I didn't sanity-check the actual counterexamples :-). But fine to merge.

No one did. Here we just want to have them out. We have an issue to check the counterexamples themselves separately.

**Pull request merged**
